### PR TITLE
do_change_locale fixed

### PR DESCRIPTION
**Pull request by garyk10**
Fixes a typo that is breaking the conditional in the do_change_locale() function

### CODE REVIEW

**Comment by XECDesign** [comment]
I don't think it's a typo. Can you provide an example where it doesn't work the way it currently is, but does with the change?

**Comment by garyk10** [comment]
sudo raspi-config nonint do_change_locale "en_US.UTF-8 UTF-8"  # fails due to "en_US.UTF-8 UTF-8 " not existing in /etc/locale.gen
sudo raspi-config nonint do_change_locale "en_US.UTF-8 "  # fails, due to two spaces (as an example; not realistic)
sudo raspi-config nonint do_change_locale "en_US.UTF-8"  # works fine because "en_US.UTF-8 " is found within the "en_US.UTF-8 UTF-8" line of /etc/locale.gen

With the extra space removed on the grep command, all of these commands would work as expected

**Comment by lurch** [comment]
It looks like the space might be there deliberately, to ensure that the `grep` only matches a single line? (and an invalid locale also doesn't partially-match a valid locale)
```
$ grep "^en_US.UTF-8 " /usr/share/i18n/SUPPORTED 
en_US.UTF-8 UTF-8
$ grep "^en_US.UTF-8" /usr/share/i18n/SUPPORTED 
en_US.UTF-8 UTF-8
$ grep "^en_US " /usr/share/i18n/SUPPORTED 
en_US ISO-8859-1
$ grep "^en_US" /usr/share/i18n/SUPPORTED 
en_US.UTF-8 UTF-8
en_US ISO-8859-1
en_US.ISO-8859-15 ISO-8859-15
$ grep "^en_US.U " /usr/share/i18n/SUPPORTED 
$ grep "^en_US.U" /usr/share/i18n/SUPPORTED 
en_US.UTF-8 UTF-8
$ grep "^en " /usr/share/i18n/SUPPORTED 
$ grep "^en" /usr/share/i18n/SUPPORTED 
en_AG UTF-8
en_AU.UTF-8 UTF-8
en_AU ISO-8859-1
en_BW.UTF-8 UTF-8
en_BW ISO-8859-1
en_CA.UTF-8 UTF-8
en_CA ISO-8859-1
en_DK.UTF-8 UTF-8
en_DK.ISO-8859-15 ISO-8859-15
en_DK ISO-8859-1
en_GB.UTF-8 UTF-8
en_GB ISO-8859-1
en_GB.ISO-8859-15 ISO-8859-15
en_HK.UTF-8 UTF-8
en_HK ISO-8859-1
en_IE.UTF-8 UTF-8
en_IE ISO-8859-1
en_IE@euro ISO-8859-15
en_IL UTF-8
en_IN UTF-8
en_NG UTF-8
en_NZ.UTF-8 UTF-8
en_NZ ISO-8859-1
en_PH.UTF-8 UTF-8
en_PH ISO-8859-1
en_SC.UTF-8 UTF-8
en_SG.UTF-8 UTF-8
en_SG ISO-8859-1
en_US.UTF-8 UTF-8
en_US ISO-8859-1
en_US.ISO-8859-15 ISO-8859-15
en_ZA.UTF-8 UTF-8
en_ZA ISO-8859-1
en_ZM UTF-8
en_ZW.UTF-8 UTF-8
en_ZW ISO-8859-1
```

Given that `sudo raspi-config nonint do_change_locale "en_US.UTF-8"` works fine, what's the requirement for `sudo raspi-config nonint do_change_locale "en_US.UTF-8 UTF-8"` to also work? :shrug:
(If you indeed wanted to fix this, I guess you'd need to modify the script to only add the extra space if the `$LOCALE` didn't already contain a space?)

(Hmmm after playing around a bit more, it looks like
```
    if ! LOCALE_LINE="$(grep -E "^$LOCALE( |$)" /usr/share/i18n/SUPPORTED)"; then
```
_might_ also work? Note the addition of the `-E`.)

**Comment by garyk10** [comment]
I don't know the intention, and it's my personal preference that a line from /usr/share/i18n/SUPPORTED could be used as the final argument for a command like
`sudo raspi-config nonint do_change_locale ${LOCALE_NAME_FULL_LINE}`
I'll leave this pull request open for a while longer, then close it if doesn't get merged.  No big deal either way.

**Comment by XECDesign** [comment]
I'd like to think that there's a good reason I've done it the way I have, but what you're saying probably makes more sense. When I get a moment, I'll double check and merge if there are no issues. It might need to work both ways in case it breaks some existing scripts.

**Comment by lurch** [comment]
My `if ! LOCALE_LINE="$(grep -E "^$LOCALE( |$)" /usr/share/i18n/SUPPORTED)"; then` suggestion above "works both ways" (subject to additional testing, of course) :slightly_smiling_face:

**Comment by XECDesign** [comment]
Yeah, that checks out. Many thanks.

**Comment by XECDesign** [comment]
Actually no, it needs a further fixup, which I'll add in a bit.

**Comment by lurch** [comment]
Ooops, what did I miss?

**Comment by XECDesign** [comment]
Not you, all of us. We currently write `$LOCALE $ENCODING` into `locale.gen` which ends up having the encoding twice if the full string is used. It should just write `$LOCALE_LINE` and we should probably support passing multiple locales.

Then what we do in `/etc/default/locale` doesn't make sense either. It just needs the LANG part of the string and it won't work if the line is commented out (as it is on my system).

I'm guessing this all worked on Buster, but this change and Bullseye changes broke it. Or maybe it's not a Bullseye issue and my setup is just weird.

Anyway, I think the approach is to use debconf to set them and then let `dpkg-reconfigure` do the actual work, which will then work in all cases. I've avoided it in the past because scripts called by `dpkg-reconfigure` don't always behave the same way in non-interactive mode. Sometimes they read the debconf values, other times they read the actual current system settings and override whatever you pre-seed. Locales' scripts seem to do the right thing as far as I can tell.

The short version is that the whole thing needs to be reworked to be useful. I'll make it so that it takes either a locale as in the previous case, or a comma-delimited list of locale and encoding strings, where the first one will be used as the default (can be None).

**Comment by lurch** [comment]
> Not you, all of us...

Ah, I was only looking at this single line in isolation, and hadn't tested it's impact on the rest of the script. That's why I said "subject to additional testing" :laughing:

**Comment by XECDesign** [comment]
Okay I lied, locale's debconf don't work in the way I'd like at all, so the original method is still the way to go.

And overcomplicating things by re-implementing half of the debconf script to support generation of multiple locales makes no sense.

https://github.com/RPi-Distro/raspi-config/commit/618a3fedce5b44ca31498389e7cdd756fed38c64
